### DISCUSSION TITLE
Character frequency

### DISCUSSION THREAD

**Discussion by iandoug** [binarybottle/engram-es#15] [Ideas]
Let me start a new thread, others are getting long.

Character frequency from analysis of cleaned and paragraphed Leipzig corpus attached. I think the licence allows to post the compiled corpus, will upload to Zenodo once I have done the bigrams and trigrams, probably only tomorrow.

Version 2:
[leipzig-spanish-char-freq.txt](https://github.com/binarybottle/engram-es/files/6915510/leipzig-spanish-char-freq.txt)

Cheers, Ian

**Comment by Lobo-Feroz** [comment]
Hey Ian, this is really awesome! 

If I have time tomorrow I'll try importing it into libreoffice calc to do some analysis.

Just by seeing the text I can see some hypothesis validated:

- The apostrophe ' is nowhere near as frequent as the quotes ". It's an order of magnitude less frequent. I don't think it deserves a key at the same level.
  - For instance, just the tilded á is more frequent than " and much, much more than '. I think the tilde or star key should have an important position in the keyboard.
- The capital Ü has a very low frequency indeed, so we don't need to make this a very easy character.

**Comment by iandoug** [comment]
Bigrams. All char pairs, I deleted those with zero.

Percentage is percentage of total count.
Make sure first column is formatted as text, or it will interpret some character pairs as formulas.

Trigrams tomorrow.

[bigrams.ods](https://github.com/binarybottle/engram-es/files/6913508/bigrams.ods)

Treat as preliminary, since corpus appears to have dropped diacriticed vowels.

**Comment by iandoug** [comment]
Have replaced char frequency with version 2 in first post above.

There was a trailing space at the end of paragraphs, which increased the 'space' count, and created the 'space enter' bigram instead of the 'period enter' bigram.

Revised bigram file will be uploaded when done ... it takes a while.

**Reply by Lobo-Feroz** [reply]
Is space usually shown in frequencies for languages? I don't recall having seen it in wikipedia et al.

Maybe for public presentations and the like we'd rather use the relative frequency for characters withouth the space. But this is easy to do with the spreadsheet.

**Reply by iandoug** [reply]
Most "frequency lists" floating around are word-based and usually intended for cryptography. 

Space frequency becomes important for people doing creative layouts that move space off the thumb.

I will also use it a few steps down the line when I create 'chained bigrams', which is artificial text that has the correct character and bigram frequency. 

Space frequency varies by language because of average word length... see this Xhosa:
Njengoko iimfanelo zesidima soluntu semvelo kunye neemfanelo zoluntu-jikelele

vs same English
Whereas recognition of the inherent dignity and of the equal and inalienable

vs same Spanish
Considerando que la libertad, la justicia y la paz en el mundo tienen por base

**Comment by binarybottle** [comment]
@iandoug & @Lobo-Feroz -- Please alert me when you've got letter frequency and letter bigram frequency lists ready for me to run in place of the crypto lists I've been relying on.  I need them in a spreadsheet or in the following format:

letters24 = ['E','A','O','S','N','I','R','L','D','C','T','U','P',
             'M','B','G','V','Q','H','F','Y','J','Z','X','K','W']
instances24 = [10912000, 10301872, 7398419, 6128524, 5838540, 5694616, 5450913, 
               4808679, 4237020, 3648080, 3643515, 3292410, 2266539, 2152702, 
               1042859, 964043, 811441, 686200, 670756, 652415, 651259, 374173, 
               344877, 153120, 64167, 26492]
bigrams = ['DE','ES','EN',...]
bigram_instances = [2088635,1882962,1850294,...]

**Comment by binarybottle** [comment]
@iandoug --

For my engram-es layout optimization code to make sense of single-letter and letter bigram frequencies, I need just the frequencies of case insensitive letters or letter bigrams -- no spaces or non-letter characters.  So I summed the cleaned-up Leipzig corpus counts for each letter across counts for that letter with whatever diacritical marks:

e	294897235
a	271738665
o	201996963
s	162205879
n	161440601
i	151438547
r	147218050
l	124544026
d	113430944
c	99562807
t	99294129
u	89773745
m	60501673
p	60057794
b	27527305
g	25968922
v	22134740
q	18664754
y	18537469
f	17869592
h	17384139
j	9976007
z	8509998
x	3992590
k	1766727
w	1021484

It will take me forever to do this for your bigrams, however.  
Could you please recompute bigram frequencies for case-insensitive letters?

**Reply by iandoug** [reply]
Attached, I uppercased each occurrence. This kept diacritic letters separate .. I don't know if you need them that way. If you want diacritic letter merged with plain letter I can rerun.
[spanish-bigrams-unicase-letters.ods](https://github.com/binarybottle/engram-es/files/6993203/spanish-bigrams-unicase-letters.ods)

Cheers, Ian

**Reply by binarybottle** [reply]
Thank you, @iandoug, but I need diacritic letters merged with plain letters.  Could you please rerun?  And if I am to include Gutenberg sources, could you provide similar output?

**Reply by binarybottle** [reply]
Would it be straightforward for you to provide onegram frequencies for diacritic-merged plain letters for both sources (Leipzig & Gutenberg) as well?

**Reply by iandoug** [reply]
Char frequencies.
There is a difference in the order between Leipzig and Gutenberg around N R I L

Spreadsheet sorted by average.
[spanish-letters-space-unicase-leipzig-gutenberg.ods](https://github.com/binarybottle/engram-es/files/7001452/spanish-letters-space-unicase-leipzig-gutenberg.ods)

**Reply by binarybottle** [reply]
Thank you!!!

**Comment by Lobo-Feroz** [comment]
Signs frequencies.

Based on @iandoug 's spanish-character-frequency-v1.ods, I have grouped the punctuation signs per key with their shifted versions.

I was getting errors when sorting by "Average", so I made a 2nd sheet in the ods for a VLOOKUP table. Probably my calc-fu is not up to par, I'm sure it could have been done in the first sheet.

The ods:

[spanish-character-frequency-v1 - signs - 20200820.ods](https://github.com/binarybottle/engram-es/files/7018122/spanish-character-frequency-v1.-.signs.-.20200820.ods)

The result:

![signs - frequencies - 20200820](https://user-images.githubusercontent.com/10242383/130156080-531cf77b-41de-40e3-9c4c-a6612b475541.png)

- The accute accent is the most used sign key (no surprises here)
- The hyphen-underscore beats the interrogation and exclamation signs (surprise here, I expected the other way around)
- Apostrophe has a very very low frequency compared to the others (no surprises here either)

@binarybottle, I hope these can help us sort the layout of the punctuation signs.

**Reply by iandoug** [reply]
> I was getting errors when sorting by "Average"

Check that the column with the characters is defined as "text", spreadsheets get confused by = or maybe + signs. It SHOULD be, which is why I supplied .ods rather than .csv.

**Reply by Lobo-Feroz** [reply]
Forced string format in the character string, numeric format for the frequency. The problem is that the calculation for the sum of frequencies is:

`=SUM(I49+I66+I67+I78+I79+I88+I89+I105+I106+I119+I120+I100+I101+I72+I73)`

(these are the sums of the frequencies for ´ á Á é É and so on...)

When I sort using the filter, by the "Average" frequency column, it turns that into this:

` =SUM(I#¡REF!+I#¡REF!+I#¡REF!+I#¡REF!+I#¡REF!+I#¡REF!+I#¡REF!+I#¡REF!+I#¡REF!+I5+I6+I#¡REF!+I#¡REF!+I#¡REF!+I#¡REF!)`

Don't know why it loses the references to the cells. At the very least, in the VLOOKUP table it works.

Edit: it's because the row numbers change when you sort, so that sum stops being valid.

**Reply by Lobo-Feroz** [reply]
@binarybottle do you have any metrics on the difficulty/score of the punctuation locations in the keyboard?

For instance, in the two center columns it's better to use the home row shifted, or the upper row unshifted?

I am thinking of an idea that is a bit different from current engram-en layout but might work.

If the most comfortable places for punctuation are the home row, index finger, center column, we could combine the comma and period in a key. That is, the period being the shifted version of the comma. Instead of the keys (normal, shifted) being ( ,; ) and ( .: ), they could be ( ,. ) and ( ;: ). 

If we place in the home row both the ( ., ) key and the ( ´ ) key we would have there the most instances of punctuation signs, much more than the rest of the signs combined.

I've set up a table with the punctuation signs not grouped by keys, but separated (except in the cases where it's unavoidable, all the accute accents go in the same key, all apostrophes in the same key as well). And then at the right the proposal of placing , and . in the same key.

This is just an idea, of course.

[spanish-character-frequency-v1 - signs - 20200820-02.ods](https://github.com/binarybottle/engram-es/files/7020987/spanish-character-frequency-v1.-.signs.-.20200820-02.ods)

![signs - frequencies - ungrouped - 20200820](https://user-images.githubusercontent.com/10242383/130222660-9acfb369-2080-4030-ac49-fd500c57426b.png)

Edit - illustration:

![sample comma period accute accent](https://user-images.githubusercontent.com/10242383/130244979-ac6fb77c-6f00-4dc2-b34f-3107f7f996b7.png)

**Reply by binarybottle** [reply]
Thank you, @Lobo-Feroz -- this is very helpful!

**Reply by binarybottle** [reply]
@iandoug -- I don't know about the relative degree of strain/discomfort in the central columns between upper keys and shift+center keys, and suspect that the effects of lateral strain and the simultaneous pressing of two keys varies across individuals. I like your idea of comma+period equalling the frequency of the accent, but wish to avoid the period being the shifted version of the comma, because I try to avoid key combinations for frequent characters.  It might serve the memory well to have the accent on the upper right center column key, since the marks are made above the letter in all cases except for the cC, and most of the letters are vowels, on the left side.

**Reply by iandoug** [reply]
Think you mean @Lobo-Feroz  ... :-)

**Reply by iandoug** [reply]
I need to stay out of this debate because I am starting to realise the benefits of punctuation being where Colemak puts it (ie basically Qwerty), but that is at odds with Engram model. Also Spanish usage seems different to English.

**Reply by Lobo-Feroz** [reply]
> @iandoug -- I don't know about the relative degree of strain/discomfort in the central columns between upper keys and shift+center keys, and suspect that the effects of lateral strain and the simultaneous pressing of two keys varies across individuals. I like your idea of comma+period equalling the frequency of the accent, but wish to avoid the period being the shifted version of the comma, because I try to avoid key combinations for frequent characters. It might serve the memory well to have the accent on the upper right center column key, since the marks are made above the letter in all cases except for the cC, and most of the letters are vowels, on the left side.

Having the diacritics key in the upper row right side (in the Y of qwerty) was also an earlier idea of mine, when you released the first version of engram-es three weeks ago with the frequencies from the cryptography website. See here: https://github.com/binarybottle/engram-es/discussions/9#discussioncomment-1092958

So that would be an OK for me.  But then we need to move the rest of the keys around. ; and : would go, I suppose in the shifted versions of , and . But we need to relocate at the very least one of ( - _ ) and ( " ' ), since there's no more room on the upper row. Or move the ¿¡ and ?! from the bottom row, though that would break symmetry.
